test/common/router/scoped_rds_test.cc

```diff
@@ -0,0 +1,296 @@
+
+    const std::string rds_config_yaml = R"EOF(
+config_source:
+  api_config_source:
+    api_type: REST
+    cluster_names:
+      - foo_rds_cluster
+    refresh_delay: { seconds: 1, nanos: 0 }
+scoped_rds_template: true
+    )EOF";
+    MessageUtil::loadFromYaml(rds_config_yaml, rds_config_);
+  }
+
+  ~ScopedRoutesTestBase() override { factory_context_.thread_local_.shutdownThread(); }
+
+  Event::SimulatedTimeSystem& timeSystem() { return time_system_; }
+
+  NiceMock<Server::Configuration::MockFactoryContext> factory_context_;
+  std::unique_ptr<ScopedRoutesConfigProviderManager> config_provider_manager_;
+  Event::SimulatedTimeSystem time_system_;
+  envoy::config::filter::network::http_connection_manager::v2::Rds rds_config_;
+};
+
+class ScopedRdsTest : public ScopedRoutesTestBase {
+protected:
+  ScopedRdsTest() = default;
+
+  ~ScopedRdsTest() override = default;
+
+  void setup() {
+    InSequence s;
+
+    Upstream::ClusterManager::ClusterInfoMap cluster_map;
+    Upstream::MockCluster cluster;
+    cluster_map.emplace("foo_cluster", cluster);
+    EXPECT_CALL(factory_context_.cluster_manager_, clusters()).WillOnce(Return(cluster_map));
+    EXPECT_CALL(cluster, info());
+    EXPECT_CALL(*cluster.info_, addedViaApi());
+    EXPECT_CALL(cluster, info());
+    EXPECT_CALL(*cluster.info_, type());
+
+    const std::string config_yaml = R"EOF(
+config_source:
+  api_config_source:
+    api_type: REST
+    cluster_names:
+      - foo_cluster
+    refresh_delay: { seconds: 1, nanos: 0 }
+scoped_routes_config_set_name: foo_scope_set
+    )EOF";
+    envoy::config::filter::network::http_connection_manager::v2::ScopedRds scoped_rds_config;
+    MessageUtil::loadFromYaml(config_yaml, scoped_rds_config);
+    provider_ = config_provider_manager_->createXdsConfigProvider(
+        scoped_rds_config, factory_context_, "foo.",
+        ScopedRoutesConfigProviderManagerOptArg(rds_config_));
+    subscription_ = &dynamic_cast<ScopedRdsConfigProvider&>(*provider_).subscription();
+  }
+
+  ScopedRdsConfigSubscription& subscription() const { return *subscription_; }
+
+  ScopedRdsConfigSubscription* subscription_;
+  Envoy::Config::ConfigProviderPtr provider_;
+};
+
+TEST_F(ScopedRdsTest, ValidateFail) {
+  setup();
+
+  ScopedRdsConfigSubscription& subscription =
+      dynamic_cast<ScopedRdsConfigProvider&>(*provider_).subscription();
+
+  // 'name' validation
+  const std::string config_yaml = R"EOF(
+name:
+scope_key_builder:
+  fragments:
+    - header_value_extractor: { name: element }
+scopes:
+  - route_configuration_name: foo_routes
+    key:
+      fragments:
+        - string_key: x-foo-key
+)EOF";
+  Protobuf::RepeatedPtrField<envoy::api::v2::ScopedRouteConfigurationsSet> resources;
+  parseScopedRoutesConfigFromYaml(*resources.Add(), config_yaml);
+  EXPECT_THROW(subscription.onConfigUpdate(resources, "1"), ProtoValidationException);
+
+  // 'scope_key_builder.fragments' validation
```

Comment:
Mind including comments about what is invalid about the configuration? Here and for other invalid config tests. Might also be valuable to verify the exception message to ensure that we're tickling the correct validation

Reply:
Done.